aten/src/ATen/native/GridSampler.cpp

```diff
@@ -0,0 +1,780 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/detail/CUDAHooksInterface.h"
+#include "ATen/native/GridSampler.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+namespace at { namespace native {
+
+using at::native::detail::GridSamplerInterpolation;
+using at::native::detail::GridSamplerPadding;
+
+namespace {
+  static inline int64_t clip_coordinates(int64_t in, int64_t clip_limit) {
+    return std::min(clip_limit - 1, std::max(in, static_cast<int64_t>(0)));
+  }
+
+  static inline bool within_bounds_2d(int64_t h, int64_t w, int64_t H, int64_t W) {
+    return h >= 0 && h < H && w >= 0 && w < W;
+  }
+
+  static inline bool within_bounds_3d(int64_t d, int64_t h, int64_t w, int64_t D, int64_t H, int64_t W) {
+    return d >= 0 && d < D && h >= 0 && h < H && w >= 0 && w < W;
+  }
+
+  template<typename scalar_t>
+  static inline void safe_add_2d(scalar_t *data, int64_t h, int64_t w,
+                                 int64_t sH, int64_t sW, int64_t H, int64_t W,
+                                 scalar_t delta) {
+    if (h >= 0 && h < H && w >= 0 && w < W) {
+      data[h * sH + w * sW] += delta;
```

Comment:
Wasn't this previously an atomicAdd?

Reply:
@SsnL maybe that explains the perf difference?

Reply:
On CPU previously it was
```
      real old_value = THTensor_(fastGet4d)(input, n, c, y, x); \
      THTensor_(fastSet4d)(input, n, c, y, x, value + old_value); \
```
So unfortunately it doesn't explain the perf difference. 

I did do some optimization in computing the offset of a value in a tensor, but that doesn't quite explain 6~10x perf difference.

Reply:
I mixed up the THCUNN and THNN.